stored_responses/stored_responses.go

```diff
@@ -75,53 +74,62 @@ func extractStoredResponsesIds(impInfo []ImpExtPrebidData,
 	impBidderReplaceImp := ImpBidderReplaceImpID{}
 
 	for index, impData := range impInfo {
-		impId, err := jsonparser.GetString(impData.Imp, "id")
+		impId := impData.ID
+		impExt, err := impData.GetImpExt()
 		if err != nil {
-			return nil, nil, nil, nil, fmt.Errorf("request.imp[%d] missing required field: \"id\"", index)
+			return nil, nil, nil, nil, err
+		}
+		impExtPrebid := impExt.GetPrebid()
+		if impExtPrebid == nil {
+			continue
 		}
 
-		if impData.ImpExtPrebid.StoredAuctionResponse != nil {
-			if len(impData.ImpExtPrebid.StoredAuctionResponse.ID) == 0 {
+		if impExtPrebid.StoredAuctionResponse != nil {
+			if len(impExtPrebid.StoredAuctionResponse.ID) == 0 {
 				return nil, nil, nil, nil, fmt.Errorf("request.imp[%d] has ext.prebid.storedauctionresponse specified, but \"id\" field is missing ", index)
 			}
-			allStoredResponseIDs = append(allStoredResponseIDs, impData.ImpExtPrebid.StoredAuctionResponse.ID)
+			allStoredResponseIDs = append(allStoredResponseIDs, impExtPrebid.StoredAuctionResponse.ID)
 
-			impAuctionResponseIDs[impId] = impData.ImpExtPrebid.StoredAuctionResponse.ID
+			impAuctionResponseIDs[impId] = impExtPrebid.StoredAuctionResponse.ID
 
 		}
-		if len(impData.ImpExtPrebid.StoredBidResponse) > 0 {
+		if len(impExtPrebid.StoredBidResponse) > 0 {
+
+			// bidders can be specified in imp.ext and in imp.ext.prebid.bidders
+			allBidderNames := make([]string, 0)
+			for bidderName := range impExtPrebid.Bidder {
+				allBidderNames = append(allBidderNames, bidderName)
+			}
+			for extData := range impExt.GetExt() {
+				// no bidders will not be processed
+				allBidderNames = append(allBidderNames, extData)
+			}
```

Comment:
It looks like we are determining all of the bidders here by looking at both `imp.ext` and `imp.ext.prebid.bidders`. I didn't think this was necessary because we have code that promotes all bidders in `imp.ext` to `imp.ext.prebid.bidders` but it appears that code is executed during the request validation step which happens just after `ProcessStoredResponses` is called. Ideally we would promote bidders to `imp.ext.prebid.bidders` as early as possible so that logic downstream like this need only look in one location but it does not appear to be a trivial refactor so perhaps it is something we can do in the future.

Nevertheless, I think that this second `for` statement may need to be sure that reserved keywords are not considered bidders. If you look at the bidder promotion code in `validateImpExt` we are using the function `isPossibleBidder` to achieve this:
```
// promote imp[].ext.BIDDER to newer imp[].ext.prebid.bidder.BIDDER location, with the later taking precedence
for k, v := range ext {
	if isPossibleBidder(k) {
		if _, exists := prebid.Bidder[k]; !exists {
			prebid.Bidder[k] = v
			prebidModified = true
		}
		delete(ext, k)
		extModified = true
	}
}
```

Reply:
Yes, this is one of those parts I got stuck with.
I tried to use `isPossibleBidder` here and ran into the circular dependency. We will need to extract it to utils maybe to use in all needed places. 
Non-bidders fields will be skipped in further processing. This function only parses stored responses ids and matches them with bidders in imp.ext. 
Later validation function will check if bidders are valid.